code/game/objects/effects/chem/puddle.dm

```diff
@@ -0,0 +1,107 @@
+
+/obj/effect/decal/cleanable/chempuddle
+	name = "puddle"
+	icon = 'icons/effects/chempuddle.dmi'
+	icon_state = "puddle"
+
+	mouse_opacity = FALSE
+
+/obj/effect/decal/cleanable/chempuddle/Initialize()
+	. = ..()
+
+	create_reagents(100)
+
+	addtimer(CALLBACK(src, .proc/evaporation_pulse), 10 SECONDS)
+
+/obj/effect/decal/cleanable/chempuddle/Destroy()
+	var/turf/T = get_turf(src)
+
+	. = ..()
+
+	for(var/obj/effect/decal/cleanable/chempuddle/CP in range(1, T))
+		CP.update_icon(TRUE)
+
+/obj/effect/decal/cleanable/chempuddle/proc/Spread(exclude=list())
+	if(reagents.total_volume < (25 * reagents.get_viscosity())) return
+	var/turf/simulated/S = get_turf(src)
+	if(!istype(S)) return
+
+	exclude |= S
+
+	var/list/possible_targets = list()
+	for(var/d in cardinal)
+		var/turf/simulated/target = get_step(src,d)
+		var/turf/simulated/origin = get_turf(src)
+		if(!(target in exclude))
```

Comment:
What is the purpose of the exclude list? It only seems to contain the source turf, which is never going to be the result of get_step(src, dir).

Reply:
Nvm I see it now, it's passed into the recursive procs.